alarm_active_table.cpp

```diff
@@ -0,0 +1,423 @@
+  /** index: alarmActiveDateAndTime */
+  netsnmp_table_helper_add_index(table_info, ASN_OCTET_STR);
+  /** index: alarmActiveIndex */
+  netsnmp_table_helper_add_index(table_info, ASN_UNSIGNED);
+
+  table_info->min_column = alarmActiveTable_COL_MIN;
+  table_info->max_column = alarmActiveTable_COL_MAX;
+
+  /*
+   * registering the table with the master agent
+   */
+  cb.get_value = alarmActiveTable_get_value;
+  cb.container = netsnmp_container_find("alarmActiveTable_primary:"
+                                        "alarmActiveTable:"
+                                        "table_container");
+  cb.can_set = 0;
+
+  DEBUGMSGTL(("initialize_table_alarmActiveTable", "Registering as table array\n"));
+
+  netsnmp_table_container_register(my_handler, table_info, &cb, cb.container, 1);
+
+  return SNMP_ERR_NOERROR;
+}
+
+/************************************************************
+ *
+ *  This routine is called for get requests to copy the data
+ *  from the context to the varbind for the request.
+ */
+int alarmActiveTable_get_value(netsnmp_request_info* request,
+                              netsnmp_index* item,
+                              netsnmp_table_request_info* table_info)
+{
+  netsnmp_variable_list* var = request->requestvb;
+  alarmActiveTable_context* context = (alarmActiveTable_context*) item;
+  switch(table_info->colnum)
+  {
+    case COLUMN_ALARMACTIVEENGINEID:
+    {
+      // Cannot obtain with SNMPv2 so use zero-length-string //
+      static std::string empty_string  = "";
+      snmp_set_var_typed_value(var, ASN_OCTET_STR,
+                               (u_char*) &empty_string,
+                               empty_string.length());
+    }
+    break;
+
+    case COLUMN_ALARMACTIVEENGINEADDRESSTYPE:
+    {
+      // This is of type InetAddressType and will either be
+      // IPv4 or IPv6
+      int ip_type;
+      ip_type = (loc_ip.find(":") == std::string::npos) ? 1 : 2;
+      snmp_set_var_typed_value(var, ASN_INTEGER,
+                               (u_char*) &ip_type,
+                               sizeof(ip_type));
+  
+    }
+    break;
+
+    case COLUMN_ALARMACTIVEENGINEADDRESS:
+    {
+      snmp_set_var_typed_value(var, ASN_OCTET_STR,
+                               (u_char*) loc_ip.c_str(),
+                               strlen(loc_ip.c_str()));
+    }
+    break;
+
+    case COLUMN_ALARMACTIVECONTEXTNAME:
+    {
+      // We do not support multiple contexts so this object
+      // will be set to the empty string
+      static std::string empty_string  = "";
+      snmp_set_var_typed_value(var, ASN_OCTET_STR,
+                               (u_char*) &empty_string,
+                               empty_string.length());
+    }
+    break;
+    
+    case COLUMN_ALARMACTIVEVARIABLES:
+    {
+      // The number of variables in alarmActiveVariableTable for
+      // this alarm. See as we have no such table this will be 0.
+      static unsigned long variables = 0;
```

Comment:
I don't think this needs to be static. (It also doesn't need to be a long - longs will be 8 bytes, but an ASN_UNSIGNED is 4, so I'd use a uint32_t to match.)